pkg/networkservice/common/refresh/client.go

```diff
@@ -74,17 +75,25 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
 	store(ctx, metadata.IsClient(t), cancel)
 
 	eventFactory := begin.FromContext(ctx)
-	timeClock := clock.FromContext(ctx)
+	clockTime := clock.FromContext(ctx)
 	// Create the afterCh *outside* the go routine.  This must be done to avoid picking up a later 'now'
 	// from mockClock in testing
-	afterCh := timeClock.After(refreshAfter)
-	go func(cancelCtx context.Context, afterCh <-chan time.Time) {
-		select {
-		case <-cancelCtx.Done():
-		case <-afterCh:
-			eventFactory.Request(begin.CancelContext(cancelCtx))
+	afterTicker := clockTime.Ticker(refreshAfter)
+	go func() {
+		defer afterTicker.Stop()
+		for {
+			select {
+			case <-cancelCtx.Done():
+				return
+			case <-afterTicker.C():
+				if err := <-eventFactory.Request(begin.CancelContext(cancelCtx)); err != nil {
+					logger.Warnf("refresh failed: %s", err.Error())
+					continue
+				}
+				return
+			}
 		}
-	}(cancelCtx, afterCh)
+	}()
```

Comment:
Please pass the variables to the go routine explicitly, rather than by closure when possible, it precludes a variety of potential issues and bugs.